File: sdk/typescript/src/onto.ts
import { sanitizeXml } from "./sanitize.js";
import { XmlNode, parseXml, elementChildren, findChild, textOf } from "./xml-parse.js";

export const ONTO_BLOCK_KINDS = new Set([
  "title", "paragraph", "caption", "equation", "citation", "item",
  "note", "footnote", "definition", "codeBlock", "annotation",
]);

export interface OntoSection {
  id: string;
  level: string;
  page: string;
  role: string;
  title: string;
}

export function xmlToOnto(xml: string): string {
  const root = parseXml(sanitizeXml(xml));
  const version = root.attrs.version ?? "";
  const metaTitleEl = findChild(root, "metadata") && findChild(findChild(root, "metadata")!, "title");
  const docTitle = metaTitleEl ? textOf(metaTitleEl) : "";

  const sections: OntoSection[] = [];
  const blocks: Record<string, string>[] = [];
  const tables: { id: string; page: string; bbox: string; caption: string; rows: string[][] }[] = [];
  const figures: Record<string, string>[] = [];
  const references: Record<string, string>[] = [];

  const walk = (node: XmlNode, section: OntoSection | null, inMetadata: boolean) => {
    const tag = node.tag;
    let current = section;
    if (tag === "section" || tag === "appendix") {
      current = {
        id: node.attrs.id ?? "",
        level: node.attrs.level ?? (tag === "appendix" ? "appendix" : ""),
        page: node.attrs.page ?? node.attrs.pageStart ?? "",
        role: node.attrs.role ?? node.attrs.semanticRole ?? (tag === "appendix" ? "appendix" : ""),
        title: "",
      };
      sections.push(current);
    } else if (tag === "metadata") {
      inMetadata = true;
    } else if (tag === "table") {
      tables.push({
        id: node.attrs.id ?? "",
        page: node.attrs.page ?? "",
        bbox: node.attrs.bbox ?? "",
        caption: (() => { const c = findChild(node, "caption"); return c ? textOf(c) : ""; })(),
        rows: elementChildren(node)
          .filter((r) => r.tag === "row")
          .map((r) => elementChildren(r).filter((c) => c.tag === "cell").map(textOf)),
      });
      return; // captions/cells belong to the table record, not the block list
    } else if (tag === "figure") {
      const image = findChild(node, "image");
      const cap = findChild(node, "caption");
      figures.push({
        id: node.attrs.id ?? "",
        page: node.attrs.page ?? "",
        bbox: node.attrs.bbox ?? "",
        caption: cap ? textOf(cap) : "",
        alt: image?.attrs.alt ?? "",
        source: node.attrs.source ?? image?.attrs.src ?? "",
      });
      return;
    } else if (tag === "reference") {
      references.push({ id: node.attrs.id ?? "", type: node.attrs.type ?? "", text: textOf(node) });
      return;
    } else if (ONTO_BLOCK_KINDS.has(tag)) {
      if (current && !inMetadata) {
        let text = textOf(node);
        if (tag === "definition" && node.attrs.term) text = `${node.attrs.term}: ${text}`;
        if (tag === "title" && !current.title) current.title = text;
        blocks.push({
          id: node.attrs.id ?? "",
          kind: tag,
          section_id: current.id,
          level: current.level,
          page: node.attrs.page ?? current.page,
          bbox: node.attrs.bbox ?? "",
          role: node.attrs.role ?? defaultOntoRole(tag),
          text,
        });
      }
    }
    for (const child of elementChildren(node)) walk(child, current, inMetadata);
  };
  walk(root, null, false);

  const lines = ["Document[1]:"];
  ontoField(lines, "version", version);
  ontoField(lines, "title", docTitle);
  ontoField(lines, "source_format", "aipdf.semantic.xml");
  lines.push("");
  ontoColumns(lines, "Sections", sections as unknown as Record<string, string>[], ["id", "level", "page", "role", "title"]);
  lines.push("");
  ontoColumns(lines, "Blocks", blocks, ["id", "kind", "section_id", "level", "page", "bbox", "role", "text"]);
  lines.push("");
  const tableRecords = tables.map((t) => ({
    id: t.id,
    page: t.page,
    bbox: t.bbox,
    caption: t.caption,
    // Pre-serialised with raw ^ / | delimiters; emitted via rawFields below.
    rows: t.rows.map((row) => row.map(ontoArrayScalar).join("^")).join("|"),
  }));
  ontoColumns(lines, "Tables", tableRecords, ["id", "page", "bbox", "caption", "rows"], new Set(["rows"]));
  lines.push("");
  ontoColumns(lines, "Figures", figures, ["id", "page", "bbox", "caption", "alt", "source"]);
  lines.push("");
  ontoColumns(lines, "References", references, ["id", "type", "text"]);
  return lines.join("\n").trimEnd();
}

export function ontoColumns(
  lines: string[],
  name: string,
  records: Record<string, string>[],
  fields: string[],
  rawFields: Set<string> = new Set(),
): void {
  lines.push(`${name}[${records.length}]:`);
  for (const field of fields) {
    const cells = records.map((record) =>
      rawFields.has(field) ? String(record[field] ?? "") : ontoScalar(record[field] ?? ""),
    );
    lines.push(`    ${field}: ${cells.join("|")}`);
  }
}

export function ontoField(lines: string[], name: string, value: string): void {
  lines.push(`    ${name}: ${ontoScalar(value)}`);
}

export function ontoScalar(value: string): string {
  return value.replace(/\s+/g, " ").trim().replace(/`/g, "'").replace(/\|/g, "/").replace(/\^/g, ";");
}

export function ontoArrayScalar(value: string): string {
  return ontoScalar(value);
}

export function defaultOntoRole(tag: string): string {
  return ({
    title: "heading",
    paragraph: "body",
    caption: "caption",
    equation: "equation",
    citation: "citation",
    item: "list-item",
    note: "note",
    footnote: "footnote",
    definition: "definition",
    codeBlock: "code",
    annotation: "annotation",
  } as Record<string, string>)[tag] ?? tag;
}

File: sdk/python/aipdf/_onto.py
from __future__ import annotations

import xml.etree.ElementTree as ET

from ._pdf import sanitize_xml
from ._markdown import text_of


def xml_to_onto(xml: str) -> str:
    root = ET.fromstring(sanitize_xml(xml))
    title_elem = root.find("./metadata/title")
    doc_title = text_of(title_elem) if title_elem is not None else ""
    sections: list[dict[str, str]] = []
    blocks: list[dict[str, str]] = []
    tables: list[dict[str, object]] = []
    figures: list[dict[str, str]] = []
    references: list[dict[str, str]] = []

    def walk(elem: ET.Element, section: dict[str, str] | None = None) -> None:
        tag = elem.tag
        current = section
        if tag in {"section", "appendix"}:
            current = {
                "id": elem.attrib.get("id", ""),
                "level": elem.attrib.get("level", "appendix" if tag == "appendix" else ""),
                "page": elem.attrib.get("page", elem.attrib.get("pageStart", "")),
                "role": elem.attrib.get("role", elem.attrib.get("semanticRole", "appendix" if tag == "appendix" else "")),
                "title": "",
            }
            sections.append(current)
        elif tag == "table":
            tables.append(
                {
                    "id": elem.attrib.get("id", ""),
                    "page": elem.attrib.get("page", ""),
                    "bbox": elem.attrib.get("bbox", ""),
                    "caption": text_of(elem.find("caption")) if elem.find("caption") is not None else "",
                    "rows": [[text_of(cell) for cell in row.findall("cell")] for row in elem.iter("row")],
                }
            )
            return
        elif tag == "figure":
            image = elem.find("image")
            figures.append(
                {
                    "id": elem.attrib.get("id", ""),
                    "page": elem.attrib.get("page", ""),
                    "bbox": elem.attrib.get("bbox", ""),
                    "caption": text_of(elem.find("caption")) if elem.find("caption") is not None else "",
                    "alt": image.attrib.get("alt", "") if image is not None else "",
                    "source": elem.attrib.get("source", image.attrib.get("src", "") if image is not None else ""),
                }
            )
            return
        elif tag == "reference":
            references.append({"id": elem.attrib.get("id", ""), "type": elem.attrib.get("type", ""), "text": text_of(elem)})
            return
        elif tag in {"title", "paragraph", "caption", "equation", "citation", "item", "note", "footnote", "definition", "codeBlock", "annotation"}:
            if current is not None and not is_metadata_child(elem, root):
                value = text_of(elem)
                if tag == "definition" and elem.attrib.get("term"):
                    value = f"{elem.attrib['term']}: {value}"
                role = elem.attrib.get("role", default_onto_role(tag))
                block = {
                    "id": elem.attrib.get("id", ""),
                    "kind": tag,
                    "section_id": current.get("id", ""),
                    "level": current.get("level", ""),
                    "page": elem.attrib.get("page", current.get("page", "")),
                    "bbox": elem.attrib.get("bbox", ""),
                    "role": role,
                    "text": value,
                }
                if tag == "title" and not current.get("title"):
                    current["title"] = value
                blocks.append(block)
        for child in elem:
            walk(child, current)

    walk(root)

    lines = ["Document[1]:"]
    _onto_field(lines, "version", root.attrib.get("version", ""))
    _onto_field(lines, "title", doc_title)
    _onto_field(lines, "source_format", "aipdf.semantic.xml")
    lines.append("")
    _onto_columns(lines, "Sections", sections, ["id", "level", "page", "role", "title"])
    lines.append("")
    _onto_columns(lines, "Blocks", blocks, ["id", "kind", "section_id", "level", "page", "bbox", "role", "text"])
    lines.append("")
    table_rows = [dict(t, rows="|".join("^".join(_onto_array_scalar(cell) for cell in row) for row in t["rows"])) for t in tables]
    # `rows` is pre-serialized with raw ^ / | delimiters and must NOT be
    # scalar-encoded again (mirrors the Rust core's column_raw handling).
    _onto_columns(lines, "Tables", table_rows, ["id", "page", "bbox", "caption", "rows"], raw_fields={"rows"})
    lines.append("")
    _onto_columns(lines, "Figures", figures, ["id", "page", "bbox", "caption", "alt", "source"])
    lines.append("")
    _onto_columns(lines, "References", references, ["id", "type", "text"])
    return "\n".join(lines).rstrip()


def _onto_columns(
    lines: list[str],
    name: str,
    records: list[dict[str, object]],
    fields: list[str],
    raw_fields: set[str] = frozenset(),
) -> None:
    lines.append(f"{name}[{len(records)}]:")
    for field in fields:
        if field in raw_fields:
            cells = (str(record.get(field, "")) for record in records)
        else:
            cells = (_onto_scalar(str(record.get(field, ""))) for record in records)
        lines.append(f"    {field}: " + "|".join(cells))


def _onto_field(lines: list[str], name: str, value: str) -> None:
    lines.append(f"    {name}: {_onto_scalar(value)}")


def _onto_scalar(value: str) -> str:
    clean = " ".join(value.split()).replace("`", "'").replace("|", "/").replace("^", ";")
    return clean


def _onto_array_scalar(value: str) -> str:
    return _onto_scalar(value)


def default_onto_role(tag: str) -> str:
    return {
        "title": "heading",
        "paragraph": "body",
        "caption": "caption",
        "equation": "equation",
        "citation": "citation",
        "item": "list-item",
        "note": "note",
        "footnote": "footnote",
        "definition": "definition",
        "codeBlock": "code",
        "annotation": "annotation",
    }.get(tag, tag)


def is_metadata_child(elem: ET.Element, root: ET.Element) -> bool:
    metadata = root.find("metadata")
    return metadata is not None and elem in list(metadata.iter())
